### PR TITLE
Z Probe Offset Calibration Menu

### PR DESCRIPTION
**Pull request by mwinters-stuff**
### Requirements

* Filling out this template is required. Pull Requests without a clear description may be closed at the maintainers' discretion.

### Description

I use a bl-touch, and find that when re-calibrating the z-offset of the probe, the manual methods are tedious and prone
to errors.

To overcome this, i have added a menu item "Cal Z Probe Offset" under "Bed Levelling"
Enabled by "PROBE_OFFSET_MENU" in Configuration_adv.h"

This starts by setting a z-offset to "PROBE_OFFSET_START" which defaults to -4 mm
Then homes the printer before presenting a menu for z-axis movement, which is used to move the nozzle to 
the desired distance from the print bed. This calculates the offset which the user can accept with "Done"

For example, in my printer the when moving the desired distance from the bed, the z reads 1.8mm, this
is calculated from -4 to give the correct offset of -2.20 mm.

### Benefits

Improves and speeds up the process of setting the z-probe offset by removing manual calculations.

### Related Issues

This is my feature request!

### CODE REVIEW

File: Marlin/src/lcd/menu/menu_probe_offset.cpp [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      calculated_z_offset = probe.offset.z + current_position.z;
+
+      if (LCD_HEIGHT >= 4) {
+        STATIC_ITEM(MSG_MOVE_Z, SS_CENTER|SS_INVERT); 
+      }
+
+      VALUE_ITEM(MSG_ZPROBE_ZOFFSET, ftostr42_52(calculated_z_offset), SS_LEFT);
+
+      SUBMENU(MSG_MOVE_1MM,  []{ _goto_manual_move_z( 1);    });
+      SUBMENU(MSG_MOVE_01MM, []{ _goto_manual_move_z( 0.1f); });
```
Comment (meteozond) [inline Marlin/src/lcd/menu/menu_probe_offset.cpp]: Probe offset is not so high value.
Don't you want to leave 0.01 step only? 
By analogy with Probe Z Offset menu I mean.

Comment (mwinters-stuff) [reply Marlin/src/lcd/menu/menu_probe_offset.cpp]: If you are configured like me, after homing Z moves to 20mm, you really dont want to move down 16 to 18mm using the .1 step with one of the usual rotary encoders fitted to most printers (like mine).

**Comment by thisiskeithb** [comment]
Your last push just merged `2.0.x` into `bugfix-2.0.x.`

**Comment by thinkyhead** [comment]
I haven't given this a test run yet. But until I get to that… Here is the procedure that would be most ideal:

- Set `probe.offset.z` to some unlikely negative value (such as -4).
- Home XYZ
  - If Z homing is to MIN and uses the probe, then we're ready to proceed.
  - If Z homing is to MAX then we need to take a probe reading in the middle.
- Move the nozzle to the middle of the bed (i.e., the safe homing XY)
  - Note: It is not safe for Marlin to _automatically_ move the nozzle below the height where the probe triggered.
- For reliable paper-test, heat up the nozzle (or prompt the user to clean it).
- Prompt the user to "move nozzle until it touches the bed. click when done."
  - A paper-test is okay here too.
  - Both coarse and fine adjustment should be available.
- Once the user clicks the encoder button, we have all the data we need to calculate the probe offset.
  - If homing to max, there's now enough data to also make a correction to the `Z_HOME_POS` (and `Z_MAX_POS`).

I outlined this procedure in some past issue and probably worked it out better there, but of course now I can't find it.

**Comment by thinkyhead** [comment]
@mwinters-stuff — I fixed your branch. To update your copy you'll need to do the following commands in the Git Console:

```
git checkout nozzle-offset-pull
git fetch origin
git reset --hard origin/nozzle-offset-pull
```

File: Marlin/src/lcd/menu/menu_probe_offset.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  START_MENU();
+  calculated_z_offset = probe.offset.z + current_position.z;
+
+  if (LCD_HEIGHT >= 4)
+    STATIC_ITEM(MSG_MOVE_Z, SS_CENTER|SS_INVERT);
+
+  VALUE_ITEM(MSG_MESH_EDIT_Z, ftostr42_52(current_position.z), SS_LEFT);
+  VALUE_ITEM(MSG_ZPROBE_ZOFFSET, ftostr42_52(calculated_z_offset), SS_LEFT);
+
+  SUBMENU(MSG_MOVE_1MM,  []{ _goto_manual_move_z( 1);    });
```
Comment (meteozond) [inline Marlin/src/lcd/menu/menu_probe_offset.cpp:91]: What about 10mm move?
```c
SUBMENU(MSG_MOVE_10MM,  []{ _goto_manual_move_z( 10);    });
```

File: Marlin/src/lcd/menu/menu_bed_leveling.cpp [hunk 3]
```diff
@@ -278,6 +278,10 @@ void menu_bed_leveling() {
     SUBMENU(MSG_LEVEL_CORNERS, _lcd_level_bed_corners);
   #endif
 
+  #if ENABLED(PROBE_OFFSET_MENU)
```
Comment (meteozond) [inline Marlin/src/lcd/menu/menu_bed_leveling.cpp]: As UBL user I can't see this menu item, don't you want to add it into `menu_ubl.cpp` too?

**Comment by thinkyhead** [comment]
This worked well in my testing, and just needed a tweak to the heading. Thanks for the contribution!

**Comment by qwewer0** [comment]
<s>Can this be used with UBL? Because Bed Leveling menu won't show up with UBL enabled.<s/>

**Comment by mwinters-stuff** [comment]
:) Once I get my 3d printer going again after moving house I will check it out - I actually keep this fork to keep my configuration files in-tree. :)

**Comment by qwewer0** [comment]
@mwinters-stuff I'm hoping that you could make https://github.com/MarlinFirmware/Marlin/issues/19650 into reality.

**Comment by laurienzu** [comment]
image
Hello, I flashed bugfix just to try this and I have this problem: the Probe wizard is not working, black LCD after homing. Is it still WIP?

**Comment by PhilippMolitor** [comment]
are there any plans on making this work for the Ender-3 V2's DWIN screen as well?

**Pull request merged**